### ISSUE TITLE
Take indexed input points, and reuse those indexes for output triangles

### ISSUE THREAD

**Issue by haxtibal** [nical/lyon#439]
Currently, lyon takes position values for input paths, and recreates position values for output triangles. For example, you feed in `builder.move_to(point(7.071068, 0.0));` and eventually receive `FillVertex { position: (7.07106, 0.0), ...}` as part of an output triangle (note how the values slightly differ). Now you can guess that `(7.071068, 0.0)` and `(7.07106, 0.0)` are actually the same points. But you can't be sure they have same identity.

Therefore it would be nice to have an API that allows passing indexes instead of points, and enable the FillTessellator to use those input indexes to describe output triangles. Something like `builder.move_to(3); /* where 3 is a index to an actual point value */` and get `FillVertex { pos_idx: 3, ... }` as output. It's the last remaining issue to make lyon shine in my application (tessellate 3D volumes to generate OpenGL assets).

Here's some [demo code](https://gist.github.com/haxtibal/799f028f0c7113422194118e7c2518d4), which should make it clearer why it's problematic for me.

[Here](https://gist.github.com/haxtibal/799f028f0c7113422194118e7c2518d4#file-lib-rs-L71), the Path API only takes points, but my input data is in the form of positions + indexes. This means, my original indexes are already "lost" at this point.
```
let mut builder = Path::builder();
builder.polygon(&self.points);
``` 

And [here](https://gist.github.com/haxtibal/799f028f0c7113422194118e7c2518d4#file-lib-rs-L83) I have to match back output triangle points to input indexes. The search by value is expensive, and it may easily fail because of the guessed delta.
```
|vertex: FillVertex| {
    let point_index_search = self.points.iter().position(
        |&point| (point.x - vertex.position.x).abs() < delta &&
        (point.y - vertex.position.y).abs() < delta);
```

The demo is producing correct results anyway, it should just be more convenient and robust to get there.

**Comment by nical** [comment]
Hi! Yes, this use case is very high on my list. I've been slowly working on a new tessellator (new-tess branch) which:
 - doesn't use fixed point numbers internally (that's the main reason why the positions come out slightly different),
 - will be able to use vertex offsets from the path data structure to allow you to build parallel vertex attribute arrays, 
 - a bunch of other motivations that aren't related to this issue.

The experimental tessellator in the new-tess branch doesn't yet support reusing path vertex offsets but I'm getting close. I haven't quite figured out some important details though, like how to model the API so that new points can be inserted during tessellation for things like self-intersections, how to decide the ids and storage of these new points and how to make it so that the user gets notified of these new points to update their other vertex attributes if any.

**Comment by haxtibal** [comment]
> Yes, this use case is very high on my list.

I really appreciate this, and I'm sure other folks will too. Rust is a natural choice when working on 3D modeling, where you have almost always indexed vertices. Having no high quality tessellator available could even mean to have to resort to another language (though no idea to which one - I only looked at earcut.js, they don't take indexed input too, and they don't have built in 3D support too, just a [simple projection to xy](https://github.com/mapbox/earcut/blob/ad016a45d0e66f1f2cc3cd0768f4e9e0f4322041/src/earcut.js#L26)).

> like how to model the API

`GeometryBuilder<FillVertex>` at the output side of `FillTessellator` is great, because it allows direct access to application data structures, instead of having to cast and copy values around. Maybe a similar approach can be found for the input side?

**Comment by nical** [comment]
A rewrite of the tessellator just made it into the master branch, with a way to to know which endpoints vertices come from when building the vertices. Release coming in a few days.

**Comment by haxtibal** [comment]
Awesome, thanks! My use case is definitely supported now, and seemingly much more. If you got some minutes to spare, may you have a look at the example below and tell if it's how you intended the feature to be used?

```
extern crate lyon;

use lyon::math::{point, Point};
use lyon::path::{EndpointId, Path};
use lyon::tessellation::geometry_builder::{BuffersBuilder, FillVertexConstructor, VertexBuffers};
use lyon::tessellation::{FillAttributes, FillOptions, FillTessellator, VertexSource};

/* I don't need input positions copied to output, and I don't need attributes like colors.
   References to input positions are enough. */
struct MyVertex {
    id: EndpointId,
}

struct Ctor;

impl FillVertexConstructor<MyVertex> for Ctor {
    fn new_vertex(&mut self, _: Point, attributes: FillAttributes) -> MyVertex {
        /* (over-)simplification: assume no new vertices are introduced during tesselation */
        let mut endpoint_id = EndpointId::INVALID;
        if let Some(vertex_source) = attributes.sources().next() {
            if let VertexSource::Endpoint { id } = vertex_source {
                endpoint_id = id;
            }
        }
        MyVertex {
            id: endpoint_id, // btw., what's the relation to attributes.as_endpoint_id()?
        }
    }
}

fn main() {
    let input = vec![
        point(0.0, 0.0),
        point(0.0, 1.0),
        point(1.0, 1.0),
        point(1.0, 0.0),
    ];
    let mut builder = Path::builder();
    builder.polygon(&input);
    let path = builder.build();

    let mut buffers: VertexBuffers<MyVertex, u16> = VertexBuffers::new();
    let mut vertex_builder = BuffersBuilder::new(&mut buffers, Ctor);
    let mut tess = FillTessellator::new();
    let result = tess.tessellate_with_ids(
        path.id_iter(),
        &path,
        None,
        &FillOptions::default(),
        &mut vertex_builder,
    );
    assert!(result.is_ok());

    /* link back endpoint ids to input position array */
    for index in buffers.indices {
        let endpoint_id = buffers.vertices[index as usize].id.to_usize();
        println!(
            "next triangle vertex = buffers[{}] = endpoint id {:?}. Position: x = {}, y = {}",
            index, endpoint_id, input[endpoint_id].x, input[endpoint_id].y
        );
    }
}
```

Output:
```
next triangle vertex = buffers[1] = endpoint id 3. Position: x = 1, y = 0
next triangle vertex = buffers[0] = endpoint id 0. Position: x = 0, y = 0
next triangle vertex = buffers[2] = endpoint id 1. Position: x = 0, y = 1
next triangle vertex = buffers[1] = endpoint id 3. Position: x = 1, y = 0
next triangle vertex = buffers[2] = endpoint id 1. Position: x = 0, y = 1
next triangle vertex = buffers[3] = endpoint id 2. Position: x = 1, y = 1
```

There are probably other ways to do the same / better. I see `PathCommand` and `FillTessellator::tessellate_events`, which may be useful when input data is not simply given as sequential positions values but involves some lookup and transformation. But didn't delve into it yet.

**Comment by nical** [comment]
Yes, that's the general idea.

> // btw., what's the relation to attributes.as_endpoint_id()?

`as_endpoint_id` is similar to what you do in your this code snipped. It iterates the sources and returns the first one that is an endpoint (if any), it's meant to be used in the simple cases where you know you don't have self-intersections nor flattened curves.


Imagine that you have a color for each endpoint stored in a separate array `colors`. You can write your vertex constructor so that borrows `colors` and look up the proper color using the endpoint ID, for example: 

```rust
struct Ctor<'l> {
    colors: &'l[Color],
}

impl FillVertexConstructor<MyVertex> for Ctor {
    fn new_vertex(&mut self, position: Point, attributes: FillAttributes) -> MyVertex {
        /* (over-)simplification: assume simple polygon with no self-intersection */
        let id = attributes.as_endpoint_id().unwrap();
        MyVertex {
            position: position.to_array(),
            color: self.colors[id.to_usize()].to_array(),
        }
    }
}

```

This way you avoid doing another pass on the whole geometry to build the real vertex.


`PathCommands` makes it easy to create you own path data structure where you decide how points are stored in memory, and basically have full control over endpoint and control point ids.

If you work with simple polygons like in the snippet, then you can avoid using `Path` and work with [`lyon::path::polygon::PolygonSlice`](https://docs.rs/lyon_path/0.15.0/lyon_path/polygon/struct.IdPolygonSlice.html) directly. You can assume the endpoint IDs to correspond to the indices of the endpoints in the polygon (whereas things are a little more complicated with `Path`).

As I write I see that I forgot to implement the id iterator for polygons that you need in `tessellate_with_ids` so I'll add it today.

You can forget about `tessellate_events` for now. It lets you tessellate an event queue that you created manually and skip building a path object but I haven't ironed out the proper API for building the event queue without a path, so it's not very useful yet.

**Comment by nical** [comment]
I added the missing iterator in lyon_path v0.15.1.

**Comment by haxtibal** [comment]
> as_endpoint_id is similar to what you do in your example. It iterates the sources and returns the first one that is an endpoint (if any), it's meant to be used in the simple cases where you know you don't have self-intersections nor flattened curves.

`let id = attributes.as_endpoint_id().unwrap();` panics on first invocation when inserted into my above example, because `as_endpoint_id()` returns `None`. See edited snippet. Is this intentional or a bug? I'd assumed it to work because there are no self-intersections nor flattened curves.

**Comment by nical** [comment]
It was a bug, I pushed a fix in lyon_tessellation 0.15.1.
